energy-dao/src/external_sc_interactions/farm_interactions.rs

```diff
@@ -0,0 +1,263 @@
+        let mut farm_state_mapper = self.farm_state(&farm_address);
+        require!(!farm_state_mapper.is_empty(), ERROR_FARM_DOES_NOT_EXIST);
+        let farm_state = farm_state_mapper.get();
+        let farm_token_id = self.get_farm_token(&farm_address);
+        let division_safety_constant = self.get_division_safety_constant(&farm_address);
+        let current_farm_position = EsdtTokenPayment::new(
+            farm_token_id,
+            farm_state.farm_token_nonce,
+            farm_state.farm_staked_value.clone(),
+        );
+
+        let claim_rewards_result = self.call_farm_claim(farm_address, current_farm_position);
+        let (new_farm_token, farm_rewards) = claim_rewards_result.into_tuple();
+
+        self.update_farm_after_claim(
+            &farm_state,
+            &mut farm_state_mapper,
+            new_farm_token,
+            farm_rewards,
+            &division_safety_constant,
+        );
+    }
+
+    #[payable("*")]
+    #[endpoint(claimUserRewards)]
+    fn claim_user_rewards(&self, farm_address: ManagedAddress) {
+        let payment = self.call_value().single_esdt();
+        let mut farm_state_mapper = self.farm_state(&farm_address);
+        require!(!farm_state_mapper.is_empty(), ERROR_FARM_DOES_NOT_EXIST);
+        let farm_state = farm_state_mapper.get();
+        let division_safety_constant = self.get_division_safety_constant(&farm_address);
+        let wrapped_farm_token_id = self.wrapped_farm_token_id().get();
+        require!(
+            wrapped_farm_token_id == payment.token_identifier,
+            ERROR_BAD_PAYMENT_TOKENS
+        );
+        let user_rewards = self.compute_user_rewards_payment(
+            &mut farm_state_mapper,
+            &payment,
+            &division_safety_constant,
+        );
+        if user_rewards.amount > 0 {
+            farm_state_mapper.update(|config| config.reward_reserve -= &user_rewards.amount);
+        }
+        let new_attributes = WrappedFarmTokenAttributes {
+            token_rps: farm_state.farm_rps,
+        };
+
+        self.send().esdt_local_burn(
+            &payment.token_identifier,
+            payment.token_nonce,
+            &payment.amount,
+        );
+        let wrapped_farm_token_id = self.wrapped_farm_token_id().get();
+        let new_farm_token =
+            self.mint_tokens(wrapped_farm_token_id, payment.amount, &new_attributes);
+
+        let mut user_payments = ManagedVec::from_single_item(new_farm_token);
+        if user_rewards.amount > 0 {
+            let wrapper_user_rewards = self.wrap_locked_token(user_rewards);
+            user_payments.push(wrapper_user_rewards);
+        }
+        let caller = self.blockchain().get_caller();
+        self.send().direct_multi(&caller, &user_payments);
+    }
+
+    #[payable("*")]
+    #[endpoint(unstakeFarm)]
+    fn unstake_farm(&self, farm_address: ManagedAddress) -> EsdtTokenPayment {
+        let payment = self.call_value().single_esdt();
```

Comment:
Feels like pretty much the entire first half is the same logic as the `claim_user_rewards` function. Could you maybe deduplicate here?

Reply:
Indeed it is similar with a few aspects, especially with the checks at the beginning, but some of those variables that are declared early on are further used down the line, so I think we would overcomplicate the code by creating a new function for the check, if we would still need to return quite a few variables.
Also, as it is right now, the `claim_user_rewards` does not also claim the farm rewards of the contract, and while both (incl unstake function) call `compute_user_rewards_payment`, the farm claim rewards function must be called prior to the aforementioned method.
If it proves to be beneficial to call the farm claim rewards function, before each `claim_user_rewards`, then we could think of deduplicating the code, and we could also completely remove the `claimFarmRewards` endpoint.